### PR TITLE
Add Sklearn containers for ONNX models

### LINKED ISSUE #233: Add Sklearn API container to ONNX models

### PR DESCRIPTION
**Pull request by interesaaat**
Closes #233.

### CODE REVIEW

**Comment by interesaaat** [comment]
This PR introduces a breaking change because we are now returning a container rather than a onnx model as result. I added an option to disable the converter, but by default I would like to have containers on because they give us a uniform API. We will have to update the onnxmltools if this PR is merged as is.

@ksaur

File: tests/test_onnxml_decision_tree_converter.py [hunk 1]
```diff
@@ -251,19 +252,20 @@ def test_random_forest_multi_random(self):
...
-        warnings.filterwarnings("ignore")
-        np.random.seed(0)
-        X = np.random.rand(1, 1)
-        X = np.array(X, dtype=np.float32)
-        y = np.random.randint(1, size=1)
-        model = RandomForestClassifier(n_estimators=5).fit(X, y)
-        model.fit(X, y)
-        self._test_classifier(X, model)
+    # # Used for small tree tests
+    # # Commenting this test for the moment because it hits a bug in ORT / ONNXMLTOOLS (https://github.com/onnx/onnxmltools/issues/415)
```
Comment (ksaur) [inline tests/test_onnxml_decision_tree_converter.py:256]: Ah, this is why the coverage drop, right? thanks for filing the[ onnx issue](https://github.com/onnx/onnxmltools/issues/415)!

File: hummingbird/ml/_container.py [hunk 2]
```diff
@@ -275,3 +281,138 @@ def score_samples(self, *inputs):
...
+    The container allows to mirror the Sklearn API.
+    """
+
+    def __init__(self, model, extra_config={}):
+        """
+        Args:
+            model: A ONNX model
+            extra_config: Some additional configuration parameter
+        """
+        if onnx_runtime_installed():
```
Comment (ksaur) [inline hummingbird/ml/_container.py:299]: i think we need another fail/`else` case here, else it will die on like 305 if this is false

File: hummingbird/ml/_container.py [hunk 3]
```diff
@@ -275,3 +281,138 @@ def score_samples(self, *inputs):
...
+
+        self.model = model
+        self._extra_config = extra_config
+
+        self._session = ort.InferenceSession(self.model.SerializeToString())
+        self._output_names = [self._session.get_outputs()[i].name for i in range(len(self._session.get_outputs()))]
+        self.input_names = [input.name for input in self._session.get_inputs()]
+
+    def _get_named_inputs(self, *inputs):
+        assert len(inputs) == len(self.input_names)
```
Comment (ksaur) [inline hummingbird/ml/_container.py:315]: can you please add a comment here? (onnx inputs are so confusing!)

File: hummingbird/ml/_topology.py [hunk 4]
```diff
@@ -100,51 +104,91 @@ def convert(topology, backend, device, extra_config={}):
...
+        # Set the device for the model.
         if device != "cpu":
-            test_data.to(device)
-        torch_model = torch.jit.trace(torch_model, test_data).eval()
+            if backend == torch.__name__ or torch.jit.__name__:
+                torch_model = torch_model.to(device)
 
-    if operator_map[operators[-1].full_name].regression:
+        # If the backend is tochscript, jit the model.
+        if backend == torch.jit.__name__:
```
Comment (ksaur) [inline hummingbird/ml/_topology.py:120]: is any of this related to onnx, or is this a separate fix?

File: hummingbird/ml/_topology.py [hunk 5]
```diff
@@ -100,51 +104,91 @@ def convert(topology, backend, device, extra_config={}):
...
+        and not operator_map[operators[idx].full_name].regression
+        and not operator_map[operators[idx].full_name].classification
+        and not operator_map[operators[idx].full_name].anomaly_detection
+        and not operator_map[operators[idx].full_name].transformer
+    ):
+        idx -= 1
+
+    assert idx >= 0, "Cannot detect container type. Please fill an issue at https://github.com/microsoft/hummingbird."
+
+    # If is a transformer, we need to check wheter there is another operator type before.
```
Comment (ksaur) [inline hummingbird/ml/_topology.py]: s/wheter/whether

I'm a big confused here still...what is going on line 156/157?  This is just "return to the end/original idx" right? so that we don't end up with something negative/nonsensical?

File: hummingbird/ml/_topology.py [hunk 6]
```diff
@@ -100,51 +104,91 @@ def convert(topology, backend, device, extra_config={}):
...
+    if operator_map[operators[idx].full_name].regression:
         # We are doing a regression task.
         if backend == torch.jit.__name__:
             container = TorchScriptSklearnContainerRegression
+        elif backend == onnx.__name__:
+            container = ONNXSklearnContainerRegression
         else:
             container = PyTorchSklearnContainerRegression
-    elif operator_map[operators[-1].full_name].anomaly_detection:
+    elif operator_map[operators[idx].full_name].anomaly_detection:
```
Comment (ksaur) [inline hummingbird/ml/_topology.py:167]: And this is guaranteed to always be in-range because of 157?

File: hummingbird/ml/operator_converters/sklearn/skl_sv.py [hunk 7]
```diff
@@ -18,7 +18,7 @@
 
 class SVC(BaseOperator, torch.nn.Module):
     def __init__(self, kernel, degree, sv, nv, a, b, gamma, coef0, classes, device):
-        super(SVC, self).__init__()
+        super(SVC, self).__init__(classification=True)
```
Comment (ksaur) [inline hummingbird/ml/operator_converters/sklearn/skl_sv.py:21]: this is due to the container change, right? do we need to change this in other places, or is this sort of a bugfix?

File: tests/test_lightgbm_converter.py [hunk 8]
```diff
@@ -280,10 +280,7 @@ def test_lightgbm_onnx(self):
         onnx_model = hummingbird.ml.convert(model, "onnx", X)
 
         # Get the predictions for the ONNX-ML model
-        session = ort.InferenceSession(onnx_model.SerializeToString())
-        output_names = [session.get_outputs()[i].name for i in range(len(session.get_outputs()))]
-        inputs = {session.get_inputs()[0].name: X}
-        onnx_pred = session.run(output_names, inputs)
+        onnx_pred = onnx_model.predict(X)
```
Comment (ksaur) [inline tests/test_lightgbm_converter.py:283]: ok, i understand now about the breaking change.  i think this is great!  (well, not the breaking part, but much more usable.)

File: tests/test_onnxml_decision_tree_converter.py (hunk 1 above)
Comment (interesaaat) [reply tests/test_onnxml_decision_tree_converter.py:256]: This plus some corner conditions I think.

File: hummingbird/ml/operator_converters/sklearn/skl_sv.py (hunk 7 above)
Comment (interesaaat) [reply hummingbird/ml/operator_converters/sklearn/skl_sv.py:21]: I think it needs to go in all classification cases. Hopefully I got all of them. For trees I set the type after the class initialization (because as the trees are implemented we cannot do at instance creation time).

File: tests/test_lightgbm_converter.py (hunk 8 above)
Comment (interesaaat) [reply tests/test_lightgbm_converter.py:283]: Yea that is what I tought too!

File: hummingbird/ml/_container.py (hunk 2 above)
Comment (interesaaat) [reply hummingbird/ml/_container.py:299]: Yes I definitely forgot this one.
